### PR TITLE
[TG-1157] New class for specialised generic class types.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This PR introduces a new class that is going to be used inside the generics concretisation code to represent a generic class after it has been instantiated.

### CODE REVIEW

File: src/java_bytecode/java_types.h [hunk 1]
```diff
@@ -374,4 +374,28 @@ void get_dependencies_from_generic_parameters(
...
+
+  ~java_specialised_generic_class_typet() {}
+};
+
+inline const bool java_is_specialised_generic_class_type(const typet &type)
+{
+  return type.get_bool(ID_C_specialised_generic_java_class);
+}
+
+inline java_specialised_generic_class_typet
```
Comment (smowton) [inline src/java_bytecode/java_types.h]: Write const and non-const versions, and return by reference like all the rest of these cast functions

File: src/java_bytecode/java_types.h [hunk 2]
```diff
@@ -374,4 +374,28 @@ void get_dependencies_from_generic_parameters(
...
+public:
+  java_specialised_generic_class_typet()
+  {
+    set(ID_C_specialised_generic_java_class, true);
+  }
+
+  ~java_specialised_generic_class_typet() {}
+};
+
+inline const bool java_is_specialised_generic_class_type(const typet &type)
```
Comment (smowton) [inline src/java_bytecode/java_types.h]: const pointless in by-value return type

File: src/util/irep_ids.def [hunk 3]
```diff
@@ -830,6 +830,7 @@ IREP_ID_ONE(integer_dereference)
 IREP_ID_TWO(C_java_generic_parameter, #java_generic_parameter)
 IREP_ID_TWO(C_java_generic_type, #java_generic_type)
 IREP_ID_TWO(C_java_generics_class_type, #java_generics_class_type)
+IREP_ID_TWO(C_specialised_generic_java_class, #specialised_class)
```
Comment (smowton) [inline src/util/irep_ids.def]: `#specialised_generic_java_class` please, it'll be more clear in irep dumps

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 4]
```diff
@@ -70,8 +80,9 @@ symbolt generate_java_generic_typet::operator()(
     pre_modification_size==after_modification_size,
     "All components in the original class should be in the new class");
 
-  const java_class_typet &new_java_class = construct_specialised_generic_type(
-    generic_class_definition, new_tag, replacement_components);
+  const java_specialised_generic_class_typet &new_java_class =
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Get by value (the return is by-value, and the reference relies on flaky lifetime-extension magic)

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 5]
```diff
@@ -41,6 +50,7 @@ symbolt generate_java_generic_typet::operator()(
 
   const irep_idt new_tag =
     build_generic_tag(existing_generic_type, generic_class_definition);
+
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: remove unrelated change

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 6]
```diff
@@ -11,6 +11,15 @@
 #include <java_bytecode/java_types.h>
 #include <java_bytecode/java_utils.h>
 
+static std::string java_type_simplify(const std::string &fqn_java_type)
+{
+  const std::string java_lang("java::java.lang");
+  const std::string package_name(java_class_to_package(fqn_java_type));
+  if(package_name==java_lang)
+    return fqn_java_type.substr(java_lang.length()+1);
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Rename "pretty-print" or something, since this produces types that should *only* be used for human readable output. Perhaps move this into type2cleanjava?

**Review by smowton — CHANGES_REQUESTED** [review]

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 7]
```diff
@@ -11,6 +11,15 @@
 #include <java_bytecode/java_types.h>
 #include <java_bytecode/java_utils.h>
 
+static std::string pretty_print_java_type(const std::string &fqn_java_type)
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generate_java_generic_type.cpp]: Could you add some Doxygen to this, in particular, with some example of the input/output of the function? I'm also wondering if this is really the best name for this function, because it seems to me its doing more than just pretty printing if I'm reading this code correctly? It's striping off the leading `java::java.lang`  - from java types that include that, so maybe the function name should make that clearer?

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 8]
```diff
@@ -11,6 +11,15 @@
...
+{
+  const std::string java_lang("java::java.lang");
+  const std::string package_name(java_class_to_package(fqn_java_type));
+  if(package_name==java_lang)
+    return fqn_java_type.substr(java_lang.length() + 1);
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generate_java_generic_type.cpp]: Rather than having a hard coded '+1' in here, is it better to just add `.` into the `java_lang` string variable?

File: src/java_bytecode/generate_java_generic_type.cpp (hunk 7 above)
Comment (chrisr-diffblue) [reply src/java_bytecode/generate_java_generic_type.cpp]: A suggestion for naming: `strip_package_from_java_class` ??? That's just a random suggestion, but it might be nice if the name was somewhat orthogonal to the exisiting `java_class_to_package` function that gets called below.

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 9]
```diff
@@ -224,24 +234,26 @@ irep_idt generate_java_generic_typet::build_generic_tag(
...
 /// \return The newly constructed class.
-java_class_typet
+java_specialised_generic_class_typet
 generate_java_generic_typet::construct_specialised_generic_type(
   const java_generic_class_typet &generic_class_definition,
   const irep_idt &new_tag,
   const struct_typet::componentst &new_components) const
 {
-  java_class_typet specialised_class = generic_class_definition;
+  java_specialised_generic_class_typet specialised_class{};
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generate_java_generic_type.cpp]: Rather than constructing an empty `java_specialised_generic_class_typet` and then filling in it's fields with a bunch of `set`s, would it be cleaner/safer to actually add a proper constructor to the class type?

File: src/java_bytecode/java_types.h [hunk 10]
```diff
@@ -374,4 +374,45 @@ void get_dependencies_from_generic_parameters(
...
+    set(ID_C_specialised_generic_java_class, true);
+  }
+
+  ~java_specialised_generic_class_typet()
+  {
+  }
+};
+
+inline bool java_is_specialised_generic_class_type(const typet &type)
+{
```
Comment (mgudemann) [inline src/java_bytecode/java_types.h]: IIRC the correspondant predicates are named `is_java_generic_FOO_BAR`, please check and change name here to adapt

File: src/java_bytecode/java_types.h [hunk 11]
```diff
@@ -374,4 +374,45 @@ void get_dependencies_from_generic_parameters(
   const typet &,
   std::set<irep_idt> &);
 
+class java_specialised_generic_class_typet : public java_generic_class_typet
```
Comment (cesaro) [inline src/java_bytecode/java_types.h]: The coding guidelines ask for American spelling for identifiers. The same applies to virtually any other hunk in the diff ;)

File: src/java_bytecode/java_types.h [hunk 12]
```diff
@@ -374,4 +374,61 @@ void get_dependencies_from_generic_parameters(
...
+  /// \param new_components: The specialised components
+  /// \return The newly constructed class.
+  java_specialized_generic_class_typet(
+    const java_generic_class_typet &generic_class_definition,
+    const irep_idt &new_tag,
+    const struct_typet::componentst &new_components)
+  {
+    set(ID_C_specialised_generic_java_class, true);
+    // We are specialising the logic - so we don't want to be marked as generic
+    set(ID_C_java_generics_class_type, false);
```
Comment (smowton) [inline src/java_bytecode/java_types.h]: If this is true, in what sense is this a child of `java_generic_class_typet`?

File: src/java_bytecode/java_types.h [hunk 13]
```diff
@@ -374,4 +374,61 @@ void get_dependencies_from_generic_parameters(
...
+{
+public:
+  /// Build the specialised version of the specific class, with the specified
+  /// parameters and name.
+  /// \param generic_class_definition: The generic class we are specialising
+  /// \param new_tag: The new name for the class (like Generic<java::Float>)
+  /// \param new_components: The specialised components
+  /// \return The newly constructed class.
+  java_specialized_generic_class_typet(
+    const java_generic_class_typet &generic_class_definition,
```
Comment (smowton) [inline src/java_bytecode/java_types.h]: Unused parameter?

File: src/java_bytecode/java_types.h [hunk 14]
```diff
@@ -374,4 +374,59 @@ void get_dependencies_from_generic_parameters(
...
+  /// \param new_components: The specialised components
+  /// \return The newly constructed class.
+  java_specialized_generic_class_typet(
+    const irep_idt &new_tag,
+    const struct_typet::componentst &new_components)
+  {
+    set(ID_C_specialized_generic_java_class, true);
+    // We are specialising the logic - so we don't want to be marked as generic
+    set(ID_C_java_generics_class_type, false);
+    set(ID_name, "java::" + id2string(new_tag));
```
Comment (smowton) [inline src/java_bytecode/java_types.h:392]: Suggest not using `name`, which is a bit vague. How about `ID_java_generic_tag` or something?

File: src/java_bytecode/java_types.h [hunk 15]
```diff
@@ -374,4 +374,59 @@ void get_dependencies_from_generic_parameters(
...
+    set(ID_base_name, id2string(new_tag));
```
Comment (smowton) [inline src/java_bytecode/java_types.h:393]: Suggest not storing this redundant info unless you really need to

File: src/java_bytecode/java_types.h [hunk 16]
```diff
@@ -374,4 +374,59 @@ void get_dependencies_from_generic_parameters(
...
+    components() = new_components;
+    const std::string &class_tag = id2string(new_tag);
+    set_tag(class_tag.substr(0, class_tag.find('<')));
```
Comment (smowton) [inline src/java_bytecode/java_types.h:396]: Are you sure multiple distinct types with matching tags is okay?

File: src/java_bytecode/java_types.h [hunk 17]
```diff
@@ -374,4 +374,59 @@ void get_dependencies_from_generic_parameters(
...
+  }
+
+  ~java_specialized_generic_class_typet()
```
Comment (smowton) [inline src/java_bytecode/java_types.h]: Might as well delete this empty destructor

File: src/java_bytecode/java_types.h (hunk 16 above)
Comment (NlightNFotis) [reply src/java_bytecode/java_types.h:396]: It is my understanding that for a type `GenericWrapper<IWrapper>` the tag of the class should always be `GenericWrapper`. That's what I have understood from discussions with @thk123 . I suggest we leave it as it is for now, as I don't have access to the specification and I don't want to make a change against the specification at this point.

File: src/java_bytecode/java_types.h (hunk 15 above)
Comment (NlightNFotis) [reply src/java_bytecode/java_types.h:393]: My understanding of the specification (and the implementation of generics thus far) is that all of the `base_name`, `name` and `tag` are needed.

File: src/java_bytecode/java_types.h (hunk 14 above)
Comment (NlightNFotis) [reply src/java_bytecode/java_types.h:392]: I think this should be staying like this. The point is it should be creating symbols that look a little bit like this:

```
  0: struct
      * name: java::reifiedArrayNB<java::java.lang.Float>
      * tag: reifiedArrayNB
      * base_name: reifiedArrayNB<java::java.lang.Float>
      * components: 
```

File: src/java_bytecode/java_types.h (hunk 16 above)
Comment (mgudemann) [reply src/java_bytecode/java_types.h:396]: @smowton this is the case we have currently, I am not aware of any problems, AFAIK the idea is still that the use of generics should be transparent for non-generics aware code

**Review by smowton — APPROVED** [review]
Still looks suspicious but I'll take it up with @thk123; go for it for now.

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 18]
```diff
@@ -224,25 +238,17 @@ irep_idt generate_java_generic_typet::build_generic_tag(
...
-/// Build the specalised version of the specific class, with the specified
+/// Build the specialised version of the specific class, with the specified
 /// parameters and name.
-/// \param generic_class_definition: The generic class we are specialising
 /// \param new_tag: The new name for the class (like Generic<java::Float>)
 /// \param new_components: The specialised components
 /// \return The newly constructed class.
-java_class_typet
+java_specialized_generic_class_typet
 generate_java_generic_typet::construct_specialised_generic_type(
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generate_java_generic_type.cpp:247]: Minor nit (non-blocking): We're getting quite a mix of `specialised` and `specialized` now, e.g. both spellings in the same function signature here. Ideally we'd clean this up, but if you don't clean it up in this PR, please raise a technical debt issue to review all the generics work and cleanup the naming in functions/comments/etc.

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me - would be nice to cleanup naming, but I recognise the mix of spellings was there before this PR, so I'm happy for this to go in as it is.
